Docs: adding games rework (#2892)
ArchipelagoMW/Archipelago: docs/adding games.md

# Adding Games

Adding a new game to Archipelago has two major parts:

* Game Modification to communicate with Archipelago server (hereafter referred to as "client")
* Archipelago Generation and Server integration plugin (hereafter referred to as "world")

This document will attempt to illustrate the bare minimum requirements and expectations of both parts of a new world
integration. As game modification wildly varies by the system and engine and has no bearing on the Archipelago protocol
it will not be detailed here.

## Client

The client is an intermediary program between the game and the Archipelago server. This can either be a direct
modification to the game, an external program, or both. This can be implemented in nearly any modern language, but it
must fulfill a few requirements in order to function as expected. The specific requirements for the game client in
order to behave as expected are as follows:

* Handle both ws and wss connections
* Detect and react when a location has been "checked" by the player by sending a network packet to the server
* Receive and parse network packets when the player receives an item from the server, and reward it to the player on
demand

>   * **Any** item can be received any number of times, up to and far surpassing those that the game might normally expect

Reviewer: I think this is more targeting the common pitfall of "I will only ever receive x amount of an item" that gets invalidated via start_inventory, plando, and item_links. Maybe a better wording would be `**Any** of your items can be received`.

>   * **Any** item can be received any number of times, up to and far surpassing those that the game might normally expect

Reviewer 2: Or if treble meant both, perhaps bold both parts: `**Any** item can be received **any number** of times`

  * Items can be sent from the server and so might not have a player or location attributed to them
* Be able to change the port for saved connection info
  * Rooms hosted on the website attempt to reserve their port, but since there are a limited amount of numbers, this
privilege can be lost, requiring the room to be moved to a new port
* Reconnect if the connection is unstable and lost while playing

> * Keep an index for items received in order to resync. Order is guaranteed.

Author: i don't really understand the confusion here. i already specify "keep an index for *items received* and a quick glance at the docs confirms it's a list which is linked and should be looked at anyways? https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/network%20protocol.md#receiveditems

> * Keep an index for items received in order to resync. Order is guaranteed.

Reviewer 2: I mean, yes, _we_ know where in the docs to take a quick glance. But someone who'd never read network_protocol before would have no way of knowing that "items received" refers to a specific piece of the AP protocol, rather than loosely referring to whatever client-side code and data structures they have to add things to as AP talks to them.

> * Keep an index for items received in order to resync. Order is guaranteed.

Reviewer 3: I agree with Ixrec here that this bullet could have some more context. "Order is guaranteed" in particular should actually state what it's talking about (what thing's order is guaranteed for what?)

* Received items that were sent to the player while they were not connected to the server
  * The player being able to complete checks and sending them when reconnecting is a good bonus but not strictly 

> required

Author: i feel like it's perfectly reasonable for a game to not be playable or able to complete checks while offline. there's no reason for it to be strictly required

* Send a status update packet alerting the server that the player has completed their goal

Libraries for most modern languages and the spec for various packets can be found in the
[network protocol](/docs/network%20protocol.md) API reference document.

## World

The world is your game integration for the Archipelago generator, webhost, and multiworld server. It contains all the
information necessary for creating the items and locations to be randomized, the logic for item placement, the 
datapackage information so other game clients can recognize your game data, and documentation. Your world must be
written as a Python package to be loaded by Archipelago. This is currently done by creating a fork of the Archipelago
repository and creating a new world package in `/worlds/`. A bare minimum world implementation must satisfy the 
following requirements:

* a folder within `/worlds/` that contains an `__init__.py`

> * A `World` subclass where you create your world and define all of its rules

Author: no just the logic of the object itself

* A mapping for items and locations defining their names and id's for clients to be able to identify them, 
`item_name_to_id` and `location_name_to_id`, respectively
* Create an item when `create_item` is called both by your code and externally
* An `options_dataclass` defining the options players have available to them.
* A "Menu" Region
* Define a non-zero amount of locations

> * Define a non-zero amount of items **equal** to the number of locations

Reviewer 4: People have been confused before about the difference between:
- putting items and locations in `item_name_to_id` and `location_name_to_id`
- Creating items and putting them in the item pool, Creating locations and putting them in regions.

This "Define" is not making that distinction clear.
If nothing else, "Create" instead of "Define" ?

* All items submitted to the multiworld must not be manually placed by the World. If you need to place specific 

> items, just place the items without ever telling the multiworld about them.

Reviewer 4: The wording here is a little unclear. It almost kind of looks like "Never use `place_locked_item`"

> 

Author: required_version is not necessary, people just keep copy pasting it. it's for breaking client changes as described in the docstring.

Notable caveats:
* The "Menu" region will always be considered the "start" for the player
* The "Menu" region is *always* considered accessible; i.e. the player is expected to always be able to return to the
start of the game from anywhere
* When submitting regions or items to the multiworld, use `append`, `extend`, or `+=`. **Do not use `=`**

> * Regions are simply containers for locations and do not have to map to real regions in your game. They can be as real
> or as abstract as you like and works for your use case.

Reviewer 2: Since "real" and "abstract" are pretty vague here, I'd prefer to give a few quick examples to get the idea across:
```suggestion
* Regions are simply containers for locations that share similar logic.
They do not have to map to concrete, physical areas within your game.
They can also represent more abstract things like tech trees, sidequest progress,
having a temporary powerup, or whatever works for your use case.
```

The base World class can be found in [AutoWorld](/worlds/AutoWorld.py). Methods available for your world to call during
generation can be found in [BaseClasses](/BaseClasses.py) and [Fill](/Fill.py). Some examples and documentation 

> regarding the API can be found in the [world api doc](/docs/world%20api.md).

Reviewer: I think a push to network protocol.md could be good, especially if we lead in with some mention of existent client libraries to get them started.

Author: ? that's on L36

# Adding Games

Adding a new game to Archipelago has two major parts:

* Game Modification to communicate with Archipelago server (hereafter referred to as "client")
* Archipelago Generation and Server integration plugin (hereafter referred to as "world")

This document will attempt to illustrate the bare minimum requirements and expectations of both parts of a new world

> integration. As game modification wildly varies by the system and engine and has no bearing on the Archipelago protocol

Reviewer 3: I'd recommend removing this "the" to match how "varies by X" is usually used. There is also a missing comma
```suggestion
integration. As game modification wildly varies by system and engine and has no bearing on the Archipelago protocol,
```

it will not be detailed here.

## Client

The client is an intermediary program between the game and the Archipelago server. This can either be a direct
modification to the game, an external program, or both. This can be implemented in nearly any modern language, but it
must fulfill a few requirements in order to function as expected. The specific requirements for the game client in

> order to behave as expected are as follows:

Reviewer 3: "in order" should be removed here to change the sentence to "The specific requirements for the game client to behave as expected are as follows:"

* Handle both ws and wss connections
* Detect and react when a location has been "checked" by the player by sending a network packet to the server
* Receive and parse network packets when the player receives an item from the server, and reward it to the player on
demand
  * **Any** of your items can be received any number of times, up to and far surpassing those that the game might
normally expect
  * Items can be sent from the server and so might not have a player or location attributed to them
* Be able to change the port for saved connection info

>   * Rooms hosted on the website attempt to reserve their port, but since there are a limited amount of numbers, this

Reviewer 3: Unfortunately, this _should_ be "number" and not "amount". I think "ports" is clearer here as well though
```suggestion
  * Rooms hosted on the website attempt to reserve their port, but since there are a limited number of ports, this
```

privilege can be lost, requiring the room to be moved to a new port
* Reconnect if the connection is unstable and lost while playing

> * Keep an index for items received in order to resync. Order is guaranteed.

Reviewer 3: None of the other lines have periods:
```suggestion
* Keep an index for items received in order to resync. Order is guaranteed
```

> * Received items that were sent to the player while they were not connected to the server

Reviewer 3: Tense matching:
```suggestion
* Receive items that were sent to the player while they were not connected to the server
```

  * The player being able to complete checks and sending them when reconnecting is a good bonus but not strictly 
required
* Send a status update packet alerting the server that the player has completed their goal

Libraries for most modern languages and the spec for various packets can be found in the
[network protocol](/docs/network%20protocol.md) API reference document.

## World

The world is your game integration for the Archipelago generator, webhost, and multiworld server. It contains all the
information necessary for creating the items and locations to be randomized, the logic for item placement, the 
datapackage information so other game clients can recognize your game data, and documentation. Your world must be
written as a Python package to be loaded by Archipelago. This is currently done by creating a fork of the Archipelago
repository and creating a new world package in `/worlds/`. A bare minimum world implementation must satisfy the 
following requirements:

> * a folder within `/worlds/` that contains an `__init__.py`

Reviewer 3: Capitalization to match style:
```suggestion
* A folder within `/worlds/` that contains an `__init__.py`
```

* A `World` subclass where you create your world and define all of its rules
* A unique game name
* For webhost documentation and behaviors, a `WebWorld` subclass that must be instantiated in the `World` class 
definition
  * the game_info doc must follow the format `{language_code}_{game_name}.md`

> * A mapping for items and locations defining their names and id's for clients to be able to identify them, 

Reviewer 3: Instead of a comma, these should be broken apart either into two sentences or with something like parentheses or —
```suggestion
* A mapping for items and locations defining their names and id's for clients to be able to identify them. These are
```

`item_name_to_id` and `location_name_to_id`, respectively
* Create an item when `create_item` is called both by your code and externally

> * An `options_dataclass` defining the options players have available to them.

Reviewer 3: ```suggestion
* An `options_dataclass` defining the options players have available to them
```

> * A "Menu" Region

Reviewer 3: ```suggestion
* A "Menu" region
```

Author: hmmm in this case I'm referring to the object. ` maybe?

> * Create a non-zero amount of locations and add them to your regions

Reviewer 3: ```suggestion
* Create a non-zero number of locations and add them to your regions
```

> * Create a non-zero amount of items **equal** to the amount of locations and add them to the multiworld itempool

Reviewer 3: ```suggestion
* Create a non-zero number of items **equal** to the number of locations and add them to the multiworld itempool
```

* All items submitted to the multiworld itempool must not be manually placed by the World. If you need to place specific 

> items, there are multiple ways to do so, but they should not be added to the multiworld itempool.

Reviewer 3: ```suggestion
items, there are multiple ways to do so, but they should not be added to the multiworld itempool
```

Notable caveats:
* The "Menu" region will always be considered the "start" for the player
* The "Menu" region is *always* considered accessible; i.e. the player is expected to always be able to return to the
start of the game from anywhere
* When submitting regions or items to the multiworld (multiworld.regions and multiworld.itempool respectively), use 
`append`, `extend`, or `+=`. **Do not use `=`**
* Regions are simply containers for locations that share similar access rules. They do not have to map to 

> concrete, physical areas within your game and can be more abstract like tech trees or a quest line.

Reviewer 3: Questline is one word
```suggestion
concrete, physical areas within your game and can be more abstract like tech trees or a questline
```

The base World class can be found in [AutoWorld](/worlds/AutoWorld.py). Methods available for your world to call during
generation can be found in [BaseClasses](/BaseClasses.py) and [Fill](/Fill.py). Some examples and documentation 
regarding the API can be found in the [world api doc](/docs/world%20api.md).

> Before publishing make sure to also check out [world maintainer.md](/docs/world%20maintainer.md).

Reviewer 3: ```suggestion
Before publishing, make sure to also check out [world maintainer.md](/docs/world%20maintainer.md).
```

# Adding Games

Adding a new game to Archipelago has two major parts:

* Game Modification to communicate with Archipelago server (hereafter referred to as "client")
* Archipelago Generation and Server integration plugin (hereafter referred to as "world")

This document will attempt to illustrate the bare minimum requirements and expectations of both parts of a new world
integration. As game modification wildly varies by system and engine, and has no bearing on the Archipelago protocol,
it will not be detailed here.

## Client

The client is an intermediary program between the game and the Archipelago server. This can either be a direct
modification to the game, an external program, or both. This can be implemented in nearly any modern language, but it
must fulfill a few requirements in order to function as expected. The specific requirements the game client must follow
to behave as expected are:

> * Handle both ws and wss connections

Reviewer 5: It might be good to expand/explain WS and WSS, you can't count on the familiarity of the acronym for someone new to modding.

* Detect and react when a location has been "checked" by the player by sending a network packet to the server
* Receive and parse network packets when the player receives an item from the server, and reward it to the player on
demand

>   * **Any** of your items can be received any number of times, up to and far surpassing those that the game might

Reviewer 5: As someone new to the project, the possibility of receiving items multiple times surprised me. I assumed the server would handle idempotency barring any connection issues. Is this due to manual resends, redundancy, connection problems? 

It would be helpful to clarify why this happens and to explain the client-developer's responsibility- making the receive code idempotent and defensive.

normally expect

>   * Items can be sent from the server and so might not have a player or location attributed to them

Reviewer 5: Another surprise. Could you explain when or why this is the case?

repository and creating a new world package in `/worlds/`. A bare minimum world implementation must satisfy the 
following requirements:

* A folder within `/worlds/` that contains an `__init__.py`
* A `World` subclass where you create your world and define all of its rules
* A unique game name
* For webhost documentation and behaviors, a `WebWorld` subclass that must be instantiated in the `World` class 
definition

>   * the game_info doc must follow the format `{language_code}_{game_name}.md`

Reviewer 2: ```suggestion
  * The game_info doc must follow the format `{language_code}_{game_name}.md`
```

> * A mapping for items and locations defining their names and id's for clients to be able to identify them. These are 

Reviewer 2: ```suggestion
* A mapping for items and locations defining their names and ids for clients to be able to identify them. These are 
```
